File: erlang/seminars/dns/src/relay.erl
-module(relay).

-export([start/0, start/1]).

-define(Server, "8.8.8.8").  %% The Google DNS server

-define(Port, 53).

start() ->
    start(?Server).

start(Server) ->
    spawn_link(fun() -> init(Server) end).


init(Server) ->
    case gen_udp:open(0, [{active, true}, binary]) of
	{ok, Socket} ->
	    relay(1, Socket, Server, []);
	Error ->
	    {error, Error}
    end.


%%% This is a bit of cheating but ok 

relay(N, Socket, Server, Msgs) ->
    receive
	{forward, From, <<Id:16, Request/binary>>} ->
	    io:format("forward message ~w/~w~n", [Id,N]),
	    gen_udp:send(Socket, Server, ?Port, <<N:16, Request/binary>>),
	    relay(N+1, Socket, Server, [{From, N, Id}|Msgs]);
	{udp, Socket, _IP, _Port, <<R:16, Reply/binary>>} ->
	    io:format("reply message ~w~n", [R]),
	    case lists:keyfind(R, 2, Msgs) of
		{From, R, Id} ->
		    From ! {reply, <<Id:16, Reply/binary>>},
		    relay(N, Socket, Server, lists:keydelete(R, 2, Msgs));
		false ->
		    relay(N, Socket, Server, Msgs)
	    end;
	update ->
	    relay:relay(N, Socket, Server, Msgs);
	stop ->
	    ok;
	Strange ->
	    io:format("strange message ~w~n", [Strange]),
	    relay(N, Socket, Server, Msgs)
    end.

    
		        

File: exercises/dns/src/relay.ex
defmodule Relay do

  @server "8.8.8.8" # The Google DNS server
  @port 53

  def start, do: start(@server)
  def start(server), do: spawn_link(fn -> init(server) end)

  defp init(server) do
    case :gen_udp.open(0, [{:active, true}, :binary]) do
      {:ok, socket} ->
        relay(1, socket, server, [])
      error ->
        {:error, error}
    end
  end

  defp relay(n, socket, server, msgs) do
    receive do
      {:forward, from, <<id :: size(16), request :: binary>>} ->
        IO.puts("Forward message: #{id}/#{n}")
        :gen_udp.send(socket, server, @port, <<n :: size(16), request :: binary>>)
        relay(n + 1, socket, server, [{from, n, id} | msgs])
      {:udp, socket, _ip, _port, <<r :: size(16), reply :: binary>>} ->
        IO.puts("Reply message: #{r}")
        case List.keyfind(msgs, r, 1) do
          {from, ^r, id} ->
            send from, {:reply, <<id :: size(16), reply :: binary>>}
            relay(n, socket, server, List.keydelete(msgs, r, 1))
          false ->
            relay(n, socket, server, msgs)
        end
      :update ->
        relay(n, socket, server, msgs)
      :stop ->
        :ok
      strange ->
        IO.puts("Strange message: #{strange}")
        relay(n, socket, server, msgs)
    end
  end
end